fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java

```diff
@@ -1452,16 +1455,101 @@ private boolean getPartitionedMVRefreshPartitions(Set<String> toRefreshPartition
+                SyncPartitionUtils.calcPotentialRefreshPartition(needRefreshMvPartitionNames, baseChangedPartitionNames,
+                        baseToMvNameRef, mvToBaseNameRef, Sets.newHashSet());
+            }
         }
         toRefreshPartitions.addAll(needRefreshMvPartitionNames);
         return true;
     }
 
+    /**
+     * According base table and materialized view's partition range, we can define those mappings from base to mv:
+     * <p>
+     * One-to-One
+     *  src:     |----|    |----|
+     *  dst:     |----|    |----|
+     * eg: base table is partitioned by one day, and mv is partition by one day
+     * </p>
+     * <p>
+     * Many-to-One
+     *  src:     |----|    |----|     |----|    |----|
+     *  dst:     |--------------|     |--------------|
+     * eg: base table is partitioned by one day, and mv is partition by date_trunc('month', dt)
+     * <p>
+     * One/Many-to-Many
+     *  src:     |----| |----| |----| |----| |----| |----|
+     *  dst:     |--------------| |--------------| |--------------|
+     * eg: base table is partitioned by three days, and mv is partition by date_trunc('month', dt)
+     *  </p>
+     *
+     *  For one-to-one or many-to-one we can trigger to refresh by materialized view's partition, but for many-to-many
+     *  we need also consider affected materialized view partitions also.
+     *
+     *  eg:
+     *  ref table's partitions:
+     *   p0:   [2023-07-27, 2023-07-30)
+     *   p1:   [2023-07-30, 2023-08-02)
+     *   p2:   [2023-08-02, 2023-08-05)
+     *  materialized view's partition:
+     *   p0:   [2023-07-01, 2023-08-01)
+     *   p1:   [2023-08-01, 2023-09-01)
+     *   p2:   [2023-09-01, 2023-10-01)
+     *
+     *  So ref table's p1 has been changed, materialized view to refresh partition: p0, p1. And when we refresh p0,p1
+     *  we also need to consider other ref table partitions(p0); otherwise, the mv's final result will lose data.
+     */
+    public boolean isCalcPotentialRefreshPartition(List<TableWithPartitions> baseChangedPartitionNames,
+                                                   Map<Table, Map<String, Range<PartitionKey>>> refBaseTableRangePartitionMap,
+                                                   Set<String> mvPartitions,
+                                                   Map<String, Range<PartitionKey>> mvPartitionNameToRangeMap) {
+        List<PartitionRange> mvSortedPartitionRanges =
+                TableWithPartitions.getSortedPartitionRanges(mvPartitionNameToRangeMap, mvPartitions);
+        for (TableWithPartitions baseTableWithPartition : baseChangedPartitionNames) {
+            Map<String, Range<PartitionKey>> baseRangePartitionMap =
+                    refBaseTableRangePartitionMap.get(baseTableWithPartition.getTable());
+            List<PartitionRange> baseSortedPartitionRanges =
+                    baseTableWithPartition.getSortedPartitionRanges(baseRangePartitionMap);
```

Comment:
That means sort the partitions many times? 
Would it be a bottleneck if table has many partitions ?

Reply:
Yep. I will refactor this later...  Now there are many places to sort `PartitionKeys`.